### DISCUSSION TITLE
Tracking variables while solving

### DISCUSSION THREAD

**Discussion by martinsgronli** [zwicker-group/py-pde#433] [Q&A] [answered]
Hi

I am solving the heat equation with temperature dependent coefficients, such as `rho` and `c`. 
`term_1 = f"3 * laplace(T)" `
`eq = PDE({"T": f"({term_1}) / ({rho} * {c})"}, bc = [bc_x_left, bc_x_right])` 
`res = eq.solve(field, t_end, dt, tracker = self.storage.tracker(dt)) `

E.g. `rho = "200 - T"` such that `rho` will be changed as the simulation runs. It will also vary in space since temperature varies in space. 

Is there a way to track the value of such coefficients in order to plot them afterwards? I need to verify that their values are actually correct.

**Comment by david-zwicker** [comment, marked as answer]
You can't directly track internal variables, but you could use a [`DataTracker`](https://py-pde.readthedocs.io/en/latest/packages/pde.trackers.trackers.html#pde.trackers.trackers.DataTracker) to track arbitrary data. I realize that it might be very useful to have an option to storage tracker, which allows to transform fields and calculated derived quantities. I'll think about it.

**Comment by david-zwicker** [comment]
I updated the `master` branch of `py-pde`, which now contains [a `transformation` argument when you initialize the tracker](https://py-pde.readthedocs.io/en/latest/packages/pde.storage.base.html#pde.storage.base.StorageBase.tracker). In your case, you should be able to use

```python
def transform(field, t):
    return FieldCollection([field, 200 - field])

term_1 = f"3 * laplace(T)" 
eq = PDE({"T": f"({term_1}) / ({rho} * {c})"}, bc = [bc_x_left, bc_x_right])
res = eq.solve(field, t_end, dt, tracker = self.storage.tracker(dt, transformation=transform)) 
```

**Reply by martinsgronli** [reply]
Thank you very much!
